File: molviewspec-ts/molviewspec/molql/expression.ts
/*
 * Copyright (c) 2018-2026 Mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author David Sehnal <david.sehnal@gmail.com>
 */

// deno-lint-ignore-file no-namespace

type Expression =
  | Expression.Literal
  | Expression.Symbol
  | Expression.Apply;

namespace Expression {
  export type Literal = string | number | boolean;
  export type Symbol = { name: string };
  export type Arguments = Expression[] | { [name: string]: Expression };
  export interface Apply {
    readonly head: Expression;
    readonly args?: Arguments;
  }

  export function Symbol(name: string): Expression.Symbol {
    return { name };
  }
  export function Apply(head: Expression, args?: Arguments): Apply {
    return args ? { head, args } : { head };
  }

  export function isArgumentsArray(e?: Arguments): e is Expression[] {
    return !!e && Array.isArray(e);
  }
  export function isArgumentsMap(e?: Arguments): e is { [name: string]: Expression } {
    return !!e && !Array.isArray(e);
  }
  export function isLiteral(e: Expression): e is Expression.Literal {
    return !isApply(e) && !isSymbol(e);
  }
  export function isApply(e: Expression): e is Expression.Apply {
    return !!e && !!(e as Expression.Apply).head && typeof e === "object";
  }
  export function isSymbol(e: Expression): e is Expression.Symbol {
    return !!e && typeof (e as any).name === "string";
  }

  /** Decide if a value has the recursive JSON shape of a MolScript expression. */
  export function is(value: unknown): value is Expression {
    if (typeof value === "string" || typeof value === "boolean") return true;
    if (typeof value === "number") return Number.isFinite(value);
    if (!value || typeof value !== "object" || Array.isArray(value)) return false;
    if (isSymbol(value as Expression)) return true;
    const apply = value as Apply;
    if (!Object.prototype.hasOwnProperty.call(apply, "head") || !is(apply.head)) return false;
    if (apply.args === undefined) return true;
    if (Array.isArray(apply.args)) return apply.args.every(is);
    return !!apply.args && typeof apply.args === "object" && Object.values(apply.args).every(is);
  }
}

export interface ValidateExpressionOptions {
  requireApplication?: boolean;
  knownSymbols?: ReadonlySet<string>;
}

export class MolQLValidationError extends Error {
  override readonly name = "MolQLValidationError";
}

/** Validate the JSON shape and application heads of a serialized MolQL expression. */
export function validateExpression(value: unknown, options: ValidateExpressionOptions = {}): Expression {
  if (!Expression.is(value)) {
    throw new MolQLValidationError("Invalid recursive MolQL expression shape");
  }
  if (options.requireApplication && !Expression.isApply(value)) {
    throw new MolQLValidationError("MolQL expression must be an application");
  }
  validateApplications(value, options.knownSymbols);
  return value;
}

function validateApplications(value: Expression, knownSymbols?: ReadonlySet<string>): void {
  if (!Expression.isApply(value)) return;
  if (!Expression.isSymbol(value.head)) {
    throw new MolQLValidationError("MolQL applications can only apply symbols");
  }
  if (knownSymbols && !knownSymbols.has(value.head.name)) {
    throw new MolQLValidationError(`MolQL symbol '${value.head.name}' is not implemented`);
  }
  if (Expression.isArgumentsArray(value.args)) {
    for (const argument of value.args) validateApplications(argument, knownSymbols);
  } else if (Expression.isArgumentsMap(value.args)) {
    for (const argument of Object.values(value.args)) validateApplications(argument, knownSymbols);
  }
}

export { Expression };

File: molviewspec/molviewspec/molql/expression.py
"""JSON-native MolQL expression types and validation helpers."""

from __future__ import annotations

import math
from typing import Any, Mapping, Sequence, TypeAlias, TypeGuard

MolQLLiteralT: TypeAlias = str | int | float | bool
MolQLSymbolT: TypeAlias = dict[str, str]
MolQLArgumentsT: TypeAlias = list["MolQLExpressionT"] | dict[str, "MolQLExpressionT"]
MolQLApplyT: TypeAlias = dict[str, Any]
MolQLExpressionT: TypeAlias = MolQLLiteralT | MolQLSymbolT | MolQLApplyT


class MolQLValidationError(ValueError):
    """Raised when a value is not a valid serialized MolQL expression."""


def symbol(name: str) -> MolQLSymbolT:
    """Create a serialized MolQL symbol."""

    if not name:
        raise MolQLValidationError("MolQL symbol names must not be empty")
    return {"name": name}


def apply(head: MolQLExpressionT, args: MolQLArgumentsT | None = None) -> MolQLApplyT:
    """Create a serialized MolQL application."""

    result: MolQLApplyT = {"head": head}
    if args is not None:
        result["args"] = args
    return result


def is_symbol(value: object) -> TypeGuard[MolQLSymbolT]:
    return isinstance(value, dict) and isinstance(value.get("name"), str)


def is_apply(value: object) -> TypeGuard[MolQLApplyT]:
    return isinstance(value, dict) and "head" in value


def is_expression(value: object) -> TypeGuard[MolQLExpressionT]:
    """Return whether *value* has the recursive JSON shape of MolQL."""

    if isinstance(value, bool) or isinstance(value, str):
        return True
    if isinstance(value, int):
        return True
    if isinstance(value, float):
        return math.isfinite(value)
    if is_symbol(value):
        return True
    if not is_apply(value) or not is_expression(value["head"]):
        return False
    if "args" not in value:
        return True
    args = value["args"]
    if isinstance(args, list):
        return all(is_expression(arg) for arg in args)
    if isinstance(args, dict):
        return all(isinstance(key, str) and is_expression(arg) for key, arg in args.items())
    return False


def validate_expression(
    value: object,
    *,
    require_application: bool = False,
    known_symbols: set[str] | None = None,
) -> MolQLExpressionT:
    """Validate a serialized MolQL expression and return it unchanged."""

    if not is_expression(value):
        raise MolQLValidationError("Invalid recursive MolQL expression shape")
    if require_application and not is_apply(value):
        raise MolQLValidationError("MolQL expression must be a MolScript application")
    _validate_applications(value, known_symbols)
    return value


def _validate_applications(value: MolQLExpressionT, known_symbols: set[str] | None) -> None:
    if not is_apply(value):
        return
    head = value["head"]
    if not is_symbol(head):
        raise MolQLValidationError("MolQL applications can only apply symbols")
    if known_symbols is not None and head["name"] not in known_symbols:
        raise MolQLValidationError(f"MolQL symbol '{head['name']}' is not implemented")
    args = value.get("args")
    if isinstance(args, Sequence) and not isinstance(args, (str, bytes)):
        for arg in args:
            _validate_applications(arg, known_symbols)
    elif isinstance(args, Mapping):
        for arg in args.values():
            _validate_applications(arg, known_symbols)
